src/nvim/winfloat.c

```diff
@@ -343,10 +352,158 @@ bool win_float_valid(const win_T *win)
+///
+/// @return true if parsing is successful, otherwise false.
+static bool parse_opt_border(WinConfig *config, char *dup_val, size_t len, Error *err)
+{
+  Object style = CSTR_AS_OBJ(dup_val);
+  parse_border_style(style, config, err);
+  api_free_object(style);
+  if (ERROR_SET(err)) {
+    return false;
+  }
+  int border_attr = syn_name2attr("FloatBorder");
+  for (int i = 0; i < 8; i++) {
+    config->border_attr[i] = config->border_hl_ids[i]
+                             ? hl_get_ui_attr(0, HLF_BORDER, config->border_hl_ids[i], false)
+                             : border_attr;
+  }
+  return true;
+}
+
+/// Parses numeric keys for 'height' and 'width' options and updates WinConfig.
+///
+/// @param fconfig Configuration storage.
+/// @param dig Digits representing the numeric value.
+/// @param len Length of the digits.
+/// @param err Pointer to the Error structure for error handling.
+///
+/// @return true if parsing is successful, otherwise false.
+static bool parse_opt_dig_key(WinConfig *config, char *dig, size_t len, Error *err)
+{
+  char *start = dig;
+  int val = getdigits_int(&dig, false, 0);
+  if (dig == start) {
+    return false;
+  }
+  if (len == 6) {
+    config->width = val;
+  } else {
+    config->height = val;
+  }
+  return true;
+}
+
+/// Parses the 'align' option for floating window configuration.
+///
+/// @param config Configuration storage.
+/// @param dup_val Value text to parse ("item" or "menu").
+/// @param len Length of the key (should be 6 for "align:").
+/// @param err Pointer to the Error structure for error handling.
+///
+/// @return true if parsing is successful, otherwise false.
+static bool parse_opt_align(WinConfig *config, char *val, size_t len, Error *err)
+{
+  // TODO(glepnir): support this by adding completepopup
+  bool is_item = (len == 4) && strncmp(val, "item", 4) == 0;
+  bool is_menu = (len == 4) && strncmp(val, "menu", 4) == 0;
+
+  if (!is_item && !is_menu) {
+    // Invalid align value
+    api_set_error(err, kErrorTypeValidation, "Invalid align value. Expected 'item' or 'menu'");
+    return false;
+  }
+
+  return true;
+}
+
+/// Parses options for configuring floating windows for completion popups or preview popups.
+/// Supports setting border style, title, title position, footer, footer position, height, and width.
+/// Only processes height and width options if `preview` is true.
+///
+/// @param fconfig The floating window configuration to modify.
+/// @param preview Indicates if the configuration is for a preview popup.
+///
+/// @return True if options are successfully parsed, otherwise false.
+bool parse_float_option(WinConfig *config)
+{
+  char *p = p_pvp;
+  Error err = ERROR_INIT;
+
+  struct {
+    char *key;
+    bool (*parser_func)(WinConfig *, char *, size_t, Error *);
+  } parsers[] = {
+    { "border:", parse_opt_border },
+    { "height:", parse_opt_dig_key },
+    { "width:", parse_opt_dig_key },
+    { "align:", parse_opt_align },
```

Comment:
was this copied from vim? or generated by AI? :) we have lots of similar code like `parse_winborder`, `parse_win_config`, `parse_border_style`. why do we need even more ? 

and if it's really needed, the name should make it more clear that this is specifically for 'previewpopup'. The name `parse_float_option` is very unclear. A clearer name would be `parse_previewpopu_option`.

Reply:
code in popupwin.c handles this option in Vim. Here's how I've modified it to make it clearer.

https://github.com/vim/vim/blob/68ee1cf7de36bdd6e642807c8beda751112aaab8/src/popupwin.c#L1778-L1904